linked_list.rb

```diff
@@ -19,77 +19,232 @@ def initialize
-    puts "Not implemented"
+    current = @head
+    size = 0
+    until size == n do
+      size += 1
+      current = current.next
+    end
+    return current.data
   end
 
   # method to insert a new node with specific data value, assuming the linked
   # list is sorted in ascending order
   def insert_ascending(value)
-    puts "Not implemented"
+    node = Node.new(value)
+    current = @head
+    # puts "helloooo #{current.data}"
+    # puts "Value: #{value}"
+    inserted = false
+    while current != nil && !inserted do
+      # puts "Count: #{count}"
+      if current.data < value
+        # puts "Current.data: #{current.data}"
+        if current.next == nil
+          # puts "Reached the end"
+          current.next = node
+          inserted = true
+        elsif current.next.data > value
+          # puts "Found the magic spot"
+          node.next = current.next
+          current.next = node
+          inserted = true
+        else
+          current = current.next
+        end
+      else
+        current = current.next
+      end
+    end
+
+    # puts "Not implemented"
   end
 
   # method to print all the values in the linked list
   def visit
-    puts "Not implemented"
+    current = @head
+    while current != nil do
+        print "#{current.data}  "
+        current = current.next
+    end
+    puts
+    # puts "Not implemented"
   end
 
   # method to delete the first node found with specified value
   def delete(value)
-    puts "Not implemented"
+    return if @head == nil
+    if @head.data == value
+      @head = @head.next
+    else
+      current = @head
+      while current.next != nil do
+        if current.next.data == value
+            current.next = current.next.next
+        else
+          current = current.next
+        end
+      end
+    end
+
   end
 
   # method to reverse the singly linked list
   # note: the nodes should be moved and not just the values in the nodes
   def reverse
-    puts "Not implemented"
+    current = @head
+    previous = nil
+
+    until current == nil do
+      temp = current.next
+      current.next = previous
+      previous = current
+      current = temp
+    end
+    @head = previous
+    # puts "Not implemented"
   end
 
   ## Advanced Exercises
   # returns the value at the middle element in the singly linked list
   def find_middle_value
-    puts "Not implemented"
+    return if @head == nil
+    return @head.data if @head.next == nil
+
+    slow = @head
+    fast = @head.next
+
+    while fast != nil do
+      slow = slow.next
+      fast = fast.next
+      if fast != nil
+        fast = fast.next
+      end
+    end
+    return slow.data
+    # puts "Not implemented"
   end
 
   # find the nth node from the end and return its value
   # assume indexing starts at 0 while counting to n
   def find_nth_from_end(n)
-    puts "Not implemented"
+    return if @head == nil
+    i = 0
+    current = @head
+
+    until i == n do
+      current = current.next

```

Comment:
Check to ensure that current is not nil before doing a .next on it.

Reply:
In other words, make your code work for cases where number of nodes is less than n.